src/Conversion/ONNXToKrnl/RNN/LSTM.cpp

```diff
@@ -0,0 +1,537 @@
+
+  // Prepare dimensions.
+  auto batchDimSize = dimAt(operandAdaptor.X(), 1);
+  auto inputDimSize = dimAt(operandAdaptor.X(), 2);
+  auto hiddenDimSize = dimAt(operandAdaptor.R(), 2);
+  Value hiddenDimVal =
+      emitConstantOp(rewriter, loc, rewriter.getIndexType(), hiddenDimSize);
+
+  auto elementType =
+      operandAdaptor.X().getType().cast<ShapedType>().getElementType();
+
+  // Prepare AffineMap to access bias, peepholes tensors.
+  AffineMap accessByOffsetMap;
+  {
+    AffineExpr iv = rewriter.getAffineDimExpr(0);
+    AffineExpr index = rewriter.getAffineSymbolExpr(0);
+    AffineExpr size = rewriter.getAffineSymbolExpr(1);
+    AffineExpr accessByOffsetExpr = index * size + iv;
+    accessByOffsetMap = AffineMap::get(1, 2, accessByOffsetExpr);
+  }
+
+  // Prepare constant indices.
+  SmallVector<Value, 4> constantIndices;
+  for (int i = 0; i < 8; i++)
+    constantIndices.emplace_back(
+        emitConstantOp(rewriter, loc, rewriter.getIndexType(), i));
+
+  // Equations for LSTM.
+  // it = f(Xt*(Wi^T) + Ht-1*(Ri^T) + Pi (.) Ct-1 + Wbi + Rbi)
+  // ft = f(Xt*(Wf^T) + Ht-1*(Rf^T) + Pf (.) Ct-1 + Wbf + Rbf)
+  // ct = g(Xt*(Wc^T) + Ht-1*(Rc^T) + Wbc + Rbc)
+  // Ct = ft (.) Ct-1 + it (.) ct
+  // ot = f(Xt*(Wo^T) + Ht-1*(Ro^T) + Po (.) Ct + Wbo + Rbo)
+  // Ht = ot (.) h(Ct)
+  //
+  // The following code will emit loops as follows:
+  // for b in 0 .. BatchDimSize
+  //   for h in 0 .. HiddenDimSize
+  //     for i in 0 .. InputDimSize {
+  //       compute Xt*(Wi^T), Xt*(Wo^T), Xt*(Wf^t), Xt*(Wc^T),
+  //               Ht-1*(Ri^T), Ht-1*(Ro^T), Ht-1*(Rf^t), Ht-1*(Rc^T)
+  //     }
+  //     compute it, ft, ct, Ct, ot, Ht
+
+  BuildKrnlLoop stateLoops(rewriter, loc, 2);
+  stateLoops.createDefineAndOptimizeOp();
+  stateLoops.pushBounds(0, batchDimSize);
+  stateLoops.pushBounds(0, hiddenDimSize);
+  stateLoops.createIterateOp();
+
+  rewriter.setInsertionPointToStart(stateLoops.getIterateBlock());
+  {
+    auto batchIV = stateLoops.getInductionVar(0);
+    auto hiddenIV = stateLoops.getInductionVar(1);
+
+    // IVs to access tensors.
+    // IVs for the hidden and cell state tensors.
+    SmallVector<Value, 4> hIVs, cIVs;
+    // IVs for the bias tensors for W and R.
+    SmallVector<SmallVector<Value, 4>, 4> wbIOFCIVs, rbIOFCIVs;
+    // IVs for the peepholes.
+    SmallVector<SmallVector<Value, 4>, 4> pIOFIVs;
+
+    { // Compute IVs.
+      // H :: [num_directions, batch_size, hidden_size]
+      hIVs = {directionIV, batchIV, hiddenIV};
+      // C :: [num_directions, batch_size, hidden_size]
+      cIVs = {directionIV, batchIV, hiddenIV};
+
+      // Bias [Wb[iofc], Rb[iofc]] :: [num_directions, 8*hidden_size]
+      if (hasBiasForInput) {
+        // Wb[iofc]
+        for (unsigned i = 0; i < 4; ++i) {
+          Value wHiddenIV =
+              rewriter.create<AffineApplyOp>(loc, accessByOffsetMap,
+                  ValueRange(std::vector<Value>{
+                      hiddenIV, constantIndices[i], hiddenDimVal}));
```

Comment:
It would be very helpful to comment as if we are actually doing a function-invocation. For instance, writing comments like `/*iv=*/hiddenIV, /*index=*/constantIndices[i] ...` to make affine map application clear.